src/lib.mo

```diff
@@ -0,0 +1,327 @@
+      case (#counter x, true) value := x;
+      case (_) {};
+    };
+  };
+
+  class GaugeValue(prefix_ : Text, buckets_ : ?[Nat], now : () -> Int, watermarkResetIntervalSeconds : Nat, isStable : Bool) {
+    public let prefix = prefix_;
+    public let buckets : ?[Nat] = switch (buckets_) {
+      case (null) null;
+      case (?b) ?Array.sort(b, Nat.compare);
+    };
+
+    class WatermarkTracker<T>(default : T, condition : (old : T, new : T) -> Bool, resetIntervalSeconds : Nat) {
+      var lastWatermarkTimestamp : Time.Time = 0;
+      public var value : T = default;
+      public func update(current : T) {
+        if (condition(value, current) or now() > lastWatermarkTimestamp + resetIntervalSeconds * 1_000_000_000) {
+          value := current;
+          lastWatermarkTimestamp := now();
+        };
+      };
+    };
+
+    public var count : Nat = 0;
+    public var sum : Nat = 0;
+    public var highWatermark : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new > old, watermarkResetIntervalSeconds);
+    public var lowWatermark : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new < old, watermarkResetIntervalSeconds);
+    public var lastValue : Nat = 0;
+    public let bucketValues : [var Nat] = Array.tabulateVar<Nat>(
+      switch (buckets) { case (?b) { b.size() + 1 }; case (null) { 0 } },
+      func(n) = 0,
+    );
+
+    public func update(current : Nat) {
+      count += 1;
+      sum += current;
+      highWatermark.update(current);
+      lowWatermark.update(current);
+      switch (buckets) {
+        case (null) {};
+        case (?b) {
+          var skip : Nat = 0;
+          label l for (bucketValue in b.vals()) {
+            if (current <= bucketValue) {
+              break l;
+            } else {
+              skip += 1;
+            };
+          };
+          for (i in Iter.range(skip, bucketValues.size() - 1)) {
+            bucketValues[i] += 1;
+          };
+        };
+      };
+    };
+
+    public func dump() : [(Text, Nat)] = Array.tabulate<(Text, Nat)>(
+      4 + bucketValues.size(),
+      func(n) = switch (n) {
+        case (0)(prefix # "_sum{}", sum);
+        case (1)(prefix # "_count{}", count);
+        case (2)(prefix # "_high_watermark{}", highWatermark.value);
+        case (3)(prefix # "_low_watermark{}", lowWatermark.value);
+        case (_)(
+          prefix
+          # "_bucket{le=\""
+          # (if (n < bucketValues.size() + 3) { Nat.toText(Option.get(buckets, [])[n - 4]) } else { "+Inf" })
+          # "\"}",
+          bucketValues[n - 4],
+        );
+      },
+    );
+
+    public func share() : ?StableDataItem {
+      if (not isStable) return null;
+      ? #gauge(count, sum);
+    };
+    public func unshare(data : StableDataItem) = switch (data, isStable) {
+      case (#gauge(c, s), true) {
+        count := c;
+        sum := s;
+      };
+      case (_) {};
+    };
+  };
```

Comment:
Shall we even allow gauge to be stable? Is there a use case?

Reply:
Removed ability to save gauges to stable memory completely